### PR TITLE
migrate siw-next into okta-signin-widget

### PR DESCRIPTION
**Pull request by lesterchoi-okta**
## Description:

Migrates `okta/siw-next` in as yarn workspace at `src/v3`.

Related to https://github.com/okta/okta-core/pull/66142

## PR Checklist

- [ ] Have you verified the basic functionality for this change?

- [ ] Added unit tests?

- [ ] Added e2e tests

- [ ] Did you follow our [Security Best Practices](https://oktawiki.atlassian.net/wiki/display/eng/Security+Best+practices)?

- Did you verify the change by running downstream monolith artifacts? (YES | NO | UNSURE)

### Issue:

- [OKTA-480523](https://oktainc.atlassian.net/browse/OKTA-480523)

### CODE REVIEW

File: .bacon.yml [hunk 1]
```diff
@@ -62,3 +62,32 @@ test_suites:
...
+      timeout: '15'
+      script_name: e2e
+      criteria: OPTIONAL
+      queue_name: small
+    - name: v3-publish
+      script_path: /root/okta/okta-signin-widget/src/v3/scripts
+      sort_order: '13'
+      timeout: '60'
+      script_name: publish
+      criteria: OPTIONAL
```
Comment (shawyu-okta) [inline .bacon.yml:92]: Should these stay "optional"?

File: scripts/setup.sh [hunk 2]
```diff
@@ -8,9 +8,9 @@ export AUTHJS_VERSION=""
 # Install required node version
 export REGISTRY_REPO="npm-topic"
 export REGISTRY="${ARTIFACTORY_URL}/api/npm/${REGISTRY_REPO}"
-setup_service node v12.13.0
+setup_service node v12.22.12
 # Use the cacert bundled with centos as okta root CA is self-signed and cause issues downloading from yarn
-setup_service yarn 1.21.1 /etc/pki/tls/certs/ca-bundle.crt
+setup_service yarn 1.18.0 /etc/pki/tls/certs/ca-bundle.crt
```
Comment (shawyu-okta) [inline scripts/setup.sh:13]: might be worth leaving a comment why we want 1.18 vs newer yarn versions

File: .gitignore [hunk 3]
```diff
@@ -1,26 +1,27 @@
...
 *.iml
-.idea
-src/config/config.json
+*.log
 /dist
-.yalc
-.env
-testenv
+build2
+node_modules
```
Comment (shawyu-okta) [inline .gitignore:18]: nit if these are directories, add trailing `/`

File: tsconfig.json [hunk 4]
```diff
@@ -12,6 +12,7 @@
     "allowSyntheticDefaultImports": true,
     "allowJs": true,
     "checkJs": false,
+    "skipLibCheck": true,
```
Comment (shuowu-okta) [inline tsconfig.json:15]: possible to add this config in src/v3 dir?

**Review by shuowu-okta — APPROVED** [review]
Looks great! Please also do a sanity check that no lock file change is from internal registry.

File: scripts/setup.sh (hunk 2 above)
Comment (aarongranick-okta) [inline scripts/setup.sh:13]: Is this issue with yarn (which I understand is related to adding new packages) relevant to Bacon?

File: .bacon.yml (hunk 1 above)
Comment (lesterchoi-okta) [reply .bacon.yml:92]: They're optional right now so as not to impede v2 development

File: tsconfig.json (hunk 4 above)
Comment (lesterchoi-okta) [reply tsconfig.json:15]: Type errors on `yarn build:release` at project root without `skipLibCheck=true`:
```
$ ttsc --noEmit false --declaration --emitDeclarationOnly --declarationMap
node_modules/@types/ajv-errors/index.d.ts(6,10): error TS2614: Module '"ajv"' has no exported member 'Ajv'. Did you mean to use 'import Ajv from "ajv"' instead?
node_modules/@types/eslint-scope/node_modules/@types/eslint/helpers.d.ts(1,6): error TS2300: Duplicate identifier 'Prepend'.
node_modules/@types/eslint-scope/node_modules/@types/eslint/lib/rules/index.d.ts(10,14): error TS2300: Duplicate identifier 'Rule'.
node_modules/@types/eslint/helpers.d.ts(1,6): error TS2300: Duplicate identifier 'Prepend'.
node_modules/@types/eslint/lib/rules/index.d.ts(10,14): error TS2300: Duplicate identifier 'Rule'.
>> error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
>> error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
>> Exited with code: 1.
>> Error executing child process: Error: Process exited with code 1.
Warning: Task "exec:build-types" failed. Use --force to continue.

Aborted due to warnings.


Execution Time (2022-05-13 14:37:33 UTC-4)
loading tasks          593ms  ▇ 1%
exec:clean             30.7s  ▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇ 70%
propertiesToJSON:main  500ms  ▇ 1%
exec:generate-config   709ms  ▇ 2%
exec:build-types       11.6s  ▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇▇ 26%
Total 44.1s

error Command failed with exit code 3.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

File: scripts/setup.sh (hunk 2 above)
Comment (lesterchoi-okta) [reply scripts/setup.sh:13]: Yes. Specifically it's an issue with `yarn workspace`

https://github.com/yarnpkg/yarn/issues/7807

Comment (lesterchoi-okta) [reply scripts/setup.sh:13]: Not related to Bacon in particular

File: .bacon.yml (hunk 1 above)
Comment (shawyu-okta) [reply .bacon.yml:92]: I'm OK with this (it makes sense). Is there any way to have Bacon selectively enforce certain test suites only if certain directories are touched? Might be a question for productivity team. Ideally we don't have to self-enforce on our PRs.

**Review by shawyu-okta — APPROVED** [review]

File: .bacon.yml [hunk 5]
```diff
@@ -9,7 +9,7 @@ test_suites:
     - name: verify-package
       script_path: /root/okta/okta-signin-widget/scripts
       sort_order: '2'
-      timeout: '10'
+      timeout: '15'
```
Comment (aarongranick-okta) [inline .bacon.yml:12]: why would this script need more time?

File: playground/mocks/config/responseConfig.js [hunk 6]
```diff
@@ -104,7 +104,7 @@ const idx = {
     // 'oda-enrollment-android',
     // 'mdm-enrollment',
     // 'authenticator-verification-custom-app-push',
-    // 'authenticator-enroll-custom-app-push'    
+    // 'authenticator-enroll-custom-app-push'
```
[the remark this answers is not in the document]
Comment (aarongranick-okta) [reply playground/mocks/config/responseConfig.js:107]: yes

File: .bacon.yml (hunk 5 above)
Comment (shuowu-okta) [reply .bacon.yml:12]: I think it's because all new dependencies and yarn build (prepare) time increase

Comment (shuowu-okta) [reply .bacon.yml:12]: @lesterchoi-okta maybe also increase timeout for `publish`, I remember last build almost hit the timeout limit

**Review by lesterchoi-okta — COMMENTED** [review]
Bacon is green, ready to merge this.
Re-enabled lib checks, i.e., `skipLibChecks=false` in `tsconfig.json`

File: tsconfig.json (hunk 4 above)
Comment (lesterchoi-okta) [reply tsconfig.json]: re-enabled lib checks, i.e., `skipLibChecks=false`

File: src/v3/src/index.ts [hunk 7]
```diff
@@ -18,6 +18,17 @@ if (process.env.NODE_ENV === 'development' || process.env.NODE_ENV === 'test') {
   import('./mocks/browser')
     .then(({ getWorker }) => getWorker())
     .then((worker) => worker?.start());
+
+  if (process.env.NODE_ENV === 'development') {
```
[the remark this answers is not in the document]
Comment (lesterchoi-okta) [reply src/v3/src/index.ts:22]: will address in testcafe story https://oktainc.atlassian.net/browse/OKTA-508218

**Review by shuowu-okta — APPROVED** [review]

**Pull request closed without merging**